README.md
# React Resumable JS

Creates an uploader component in React, to use with [Resumable JS.](http://www.resumablejs.com/)

[![NPM](https://nodei.co/npm/react-resumable-js.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/react-resumable-js/)

[![CircleCI](https://circleci.com/gh/Artear/ReactResumableJS.svg)](https://circleci.com/gh/Artear/ReactResumableJS) [![CircleCI](https://circleci.com/gh/Artear/ReactResumableJS.svg?style=shield)](https://circleci.com/gh/Artear/ReactResumableJS)

On file added, the upload will begin.

### Changelog:
[Changelog](changelog.md)

### Install
`npm i react-resumable-js`

### Options
- uploaderID The uploader ID. Ex: "image-upload"
- dropTargetID The dropTarget ID. Ex: "myDropTarget"
- fileAccept content type file accept on input file Ex: 'image/*'
- filetypes The allowed files extensions to upload. Ex: "['jpg', 'png']"
- maxFileSize The allowed file size for upload. Is expressed in bytes. Default is: 10240000 (10mb).
- onMaxFileSizeErrorCallback Usefull to use with the above param, and take the exception for use as you want.
- fileAddedMessage The message to print when file is added. Optional. Ex: 'Starting....'
- completedMessage The message to print when file is completely uploaded. Optional. Ex: 'Completed!'
- service The service that will receive the file. Ex: 'http://www.someurl.com/myservice/image.json'
- textLabel The label of the upload. Ex: 'What photo do you want to add?'
- previousText A Text that will be displayed before the component. Optional.
- disableDragAndDrop True to disable Drag and Drop. Enable by default.
- showFileList Show or hide the filelist of uploaded files. accept boolean value
- onUploadErrorCallback Function to call on Upload error. @returns file and message
- onFileAddedError Function to call on File Added error. @returns file and errorCount
- onFileRemoved Function to call on File Removed. @return file object
- headerObject Optional, if you need to add a headers object.
- onFileSuccess Method to call when file is upload. Usually a method to set the filename that was uploaded by the component.
- disableInput Boolean to disable or enable input file. Send true to disable, false otherwise.
- maxFiles Indicates how many files can be uploaded in a single session. Valid values are any positive integer and undefined for no limit. (Default: undefined)
- fileNameServer Indicate the fileNameServer Object if the server return an object. Ex {file:"image.jpg"}, so the fileNameServer is "file"
- tmpDir path to render the preview image on the filelist, if the tmpDir is not set the preview will be a base64encode image ( low performance ). We recommend set the tmpDir
- chunkSize The size in bytes of each uploaded chunk of data (Default: 1*1024*1024)
- simultaneousUploads Number of simultaneous uploads (Default: 1)
- fileParameterName The name of the multipart POST parameter to use for the file chunk (Default: file)
- generateUniqueIdentifier Override the function that generates unique identifiers for each file. (Default: null)
- maxFilesErrorCallback A function which displays the please upload n file(s) at a time message. (Default: displays an alert box with the message Please n one file(s) at a time.)
- startButton Boolean value to show the start button
- pauseButton Boolean value to show the pause button
- cancelButton Boolean value to show the cancel button
- forceChunkSize Boolean value to force size of a chunk. (Default: false)

### Folders
- build: last deploy build
- example: webpack dev server to run the demo
    - server: nodejs server to upload the files
- src: source code

### Example

```javascript
export default class ExampleForm extends React.Component {

  constructor(props) {
    super(props);
  }

  render() {

      return (
        <fieldset>
          <p>You can add other inputs, selects or stuff right here to complete a form.</p>
          <ReactResumableJs
            uploaderID="image-upload"
            dropTargetID="myDropTarget"
            filetypes={["jpg", "png"]}
            fileAccept="image/*"
            fileAddedMessage="Started!"
            completedMessage="Complete!"
            service="http://localhost:3000/upload"
            textLabel="Uploaded files"
            previousText="Drop to upload your media:"
            disableDragAndDrop={true}
            onFileSuccess={(file, message) => {
              console.log(file, message);
            }}
            onFileAdded={(file, resumable) => {
              resumable.upload();
            }}
            maxFiles={1}
            startButton={true}
            pauseButton={false}
            cancelButton={false}
            onStartUpload={() => {
                console.log("Start upload");
            }}
            onCancelUpload={() => {
                this.inputDisable = false;
            }}
            onPauseUpload={() =>{
                this.inputDisable = false;
            }}
            onResumeUpload={() => {
                this.inputDisable = true;
            }}
          />
        </fieldset>
      );
    }
});
```

### Demo
`npm run demo`


changelog.md
### Changelog:
- 1.1.24
    - Add forceChunkSize https://github.com/Artear/ReactResumableJS/issues/25
- 1.1.23
    - Fix and improvements https://github.com/Artear/ReactResumableJS/pull/24
- 1.1.22
    - Remove invalid import
- 1.1.21
    - Fix PropTypes
    - Upgrade React-dom
- 1.1.20
    - Fix https://github.com/Artear/ReactResumableJS/pull/22
    - Fix https://github.com/Artear/ReactResumableJS/issues/21
- 1.1.19
    - Now you can pause, cancel or resume the upload
    - added start button option
    - added pause button option
    - added cancel button option
    - added onStartUpload callback
    - added onPauseUpload callback
    - added onCancelUpload callback
- 1.1.16
    - added option maxFilesErrorCallback
- 1.1.15
    - fixed low performance with the preview image in the file list.
    - added new options.
        - fileNameServer
        - tmpDir
        - chunkSize
        - simultaneousUploads
        - fileParameterName
        - generateUniqueIdentifier
    - updated demo.
- 1.1.13
    - added class disabled on label
- 1.1.12
    - removed "capture", cause some cellphones takes this, to open the cammera, instead open file browse.
- 1.1.11
    - Added MaxFileSize value for size validation (parameter and callback)
    - Added "disabled" param for the file input
- 1.1.10
    - Uploader file input is now contained on a Label for easy styling (when you need to hide file input).
- 1.1.7
    - added uniqid to elements on filelist.
    - added option onFileRemoved.
    - added option showFileList.
    - removed message status.
    - updated example.
- 1.1.6
    - minor fixed
- 1.1.5
    - added custom dropTargetId
- 1.1.4
    - added support to video and others in preview list
    - added npm run demo command
- 1.1.3
    - fixed example
- 1.1.2
    - changed unique identifier
    - changed option onFileSuccess, now return file uploaded and server message
- 1.1.0
    - Removed property options
    - Now all options are properties ( see the example )
    - Refactor: functions to arrow functions

src/ReactResumableJs.js
/**
 * Resumable JS for React JS
 * @author Gonzalo Rubino gonzalo_rubino@artear.com || gonzalorubino@gmail.com
 * @version 1.1.0
 *
 * Creates an uploader component in React, to use with Resumable JS
 * On file added, the upload will begin.
 */

import React from "react";
import Resumablejs from "resumablejs";

export default class ReactResumableJs extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            progressBar: 0,
            messageStatus: '',
            fileList: {files: []},
            isPaused: false,
            isUploading: false
        };

        this.resumable = null;
    }

    componentDidMount = () => {

        let ResumableField = new Resumablejs({
            target: this.props.service,
            query: this.props.query || {},
            fileType: this.props.filetypes,
            maxFiles: this.props.maxFiles,
            maxFileSize: this.props.maxFileSize,
            fileTypeErrorCallback: (file, errorCount) => {
                if (typeof this.props.onFileAddedError === "function") {
                    this.props.onFileAddedError(file, errorCount);
                }
            },
            maxFileSizeErrorCallback: (file, errorCount) => {
                if (typeof  this.props.onMaxFileSizeErrorCallback === "function") {
                    this.props.onMaxFileSizeErrorCallback(file, errorCount);
                }
            },
            testMethod: this.props.testMethod || 'post',
            testChunks: this.props.testChunks || false,
            headers: this.props.headerObject || {},
            chunkSize: this.props.chunkSize,
            simultaneousUploads: this.props.simultaneousUploads,
            fileParameterName: this.props.fileParameterName,
            generateUniqueIdentifier: this.props.generateUniqueIdentifier,
            forceChunkSize: this.props.forceChunkSize
        });

        if (typeof this.props.maxFilesErrorCallback === "function") {
            ResumableField.opts.maxFilesErrorCallback = this.props.maxFilesErrorCallback;
        }

        ResumableField.assignBrowse(this.uploader);

        //Enable or Disable DragAnd Drop
        if (this.props.disableDragAndDrop === false) {
            ResumableField.assignDrop(this.dropZone);
        }

        ResumableField.on('fileAdded', (file, event) => {
            this.setState({
                messageStatus: this.props.fileAddedMessage || ' Starting upload! '
            });

            if (typeof this.props.onFileAdded === "function") {
                this.props.onFileAdded(file, this.resumable);
            } else {
                ResumableField.upload();
            }
        });

        ResumableField.on('fileSuccess', (file, fileServer) => {

            if (this.props.fileNameServer) {
                let objectServer = JSON.parse(fileServer);
                file.fileName = objectServer[this.props.fileNameServer];
            } else {
                file.fileName = fileServer;
            }

            let currentFiles = this.state.fileList.files;
            currentFiles.push(file);

            this.setState({
                fileList: {files: currentFiles},
                messageStatus: this.props.completedMessage + file.fileName || fileServer
            }, () => {
                if (typeof this.props.onFileSuccess === "function") {
                    this.props.onFileSuccess(file, fileServer);
                }
            });
        });

        ResumableField.on('progress', () => {


            this.setState({
                isUploading: ResumableField.isUploading()
            });

            if ((ResumableField.progress() * 100) < 100) {
                this.setState({
                    messageStatus: parseInt(ResumableField.progress() * 100, 10) + '%',
                    progressBar: ResumableField.progress() * 100
                });
            } else {
                setTimeout(() => {
                    this.setState({
                        progressBar: 0
                    })
                }, 1000);
            }

        });

        ResumableField.on('fileError', (file, errorCount) => {
            this.props.onUploadErrorCallback(file, errorCount);
        });

        this.resumable = ResumableField;
    };

    removeFile = (event, file, index) => {

        event.preventDefault();

        let currentFileList = this.state.fileList.files;
        delete currentFileList[index];

        this.setState({
            fileList: {files: currentFileList}
        });

        this.props.onFileRemoved(file);
        this.resumable.removeFile(file);
    };

    createFileList = () => {

        let markup = this.state.fileList.files.map((file, index) => {

            let uniqID = this.props.uploaderID + '-' + index;
            let originFile = file.file;
            let media = '';

            if (file.file.type.indexOf('video') > -1) {
                media = <label className="video">{originFile.name}</label>;
                return <li className="thumbnail" key={uniqID}>
                    <label id={"media_" + uniqID}>{media}</label>
                    <a onClick={(event) => this.removeFile(event, file, index)} href="#">[X]</a>
                </li>;
            }
            else if (file.file.type.indexOf('image') > -1) if (this.props.tmpDir !== "") {
                let src = this.props.tmpDir + file.fileName;
                media = <img className="image" width="80" src={src} alt=""/>;
                return <li className="thumbnail" key={uniqID}>
                    <label id={"media_" + uniqID}>{media}</label>
                    <a onClick={(event) => this.removeFile(event, file, index)} href="#">[X]</a>
                </li>;

            } else {
                let fileReader = new FileReader();
                fileReader.readAsDataURL(originFile);
                fileReader.onload = (event) => {
                    media = '<img class="image" width="80" src="' + event.target.result + '"/>';
                    document.querySelector("#media_" + uniqID).innerHTML = media;
                };
                return <li className="thumbnail" key={uniqID}>
                    <label id={"media_" + uniqID}/>
                    <a onClick={(event) => this.removeFile(event, file, index)} href="#">[X]</a>
                </li>;
            } else {
                media = <label className="document">{originFile.name}</label>;
                return <li className="thumbnail" key={uniqID}>
                    <label id={"media_" + uniqID}>{media}</label>
                    <a onClick={(event) => this.removeFile(event, file, index)} href="#">[X]</a>
                </li>;
            }
        });

        return <ul id={"items-" + this.props.uploaderID}>{markup}</ul>;
    };

    cancelUpload = () => {
        this.resumable.cancel();

        this.setState({
            fileList: {files: []}
        });

        this.props.onCancelUpload();
    };

    pauseUpload = () => {
        if (!this.state.isPaused) {

            this.resumable.pause();
            this.setState({
                isPaused: true
            });
            this.props.onPauseUpload();
        } else {

            this.resumable.upload();
            this.setState({
                isPaused: false
            });
            this.props.onResumeUpload();
        }
    };

    startUpload = () => {
        this.resumable.upload();
        this.setState({
            isPaused: false
        });
        this.props.onStartUpload();
    };

    render() {

        let fileList = null;
        if (this.props.showFileList) {
            fileList = <div className="resumable-list">{this.createFileList()}</div>;
        }

        let previousText = null;
        if (this.props.previousText) {
            if (typeof this.props.previousText ==="string") previousText = <p>{this.props.previousText}</p>
            else previousText = this.props.previousText
        }

        let textLabel = null;
        if (this.props.textLabel) {
            textLabel = this.props.textLabel;

        }

        let startButton = null;
        if (this.props.startButton) {
            if (typeof this.props.startButton ==="string" || typeof this.props.startButton ==="boolean" ) startButton = <label>
                <button disabled={this.state.isUploading} className="btn start" onClick={this.startUpload}>{this.props.startButton && "upload"}
                </button>
            </label>;
            else startButton =this.props.startButton
        }

        let cancelButton = null;
        if (this.props.cancelButton) {
            if (typeof this.props.cancelButton ===  "string" || typeof this.props.cancelButton ===  "boolean")cancelButton = <label>
                <button disabled={!this.state.isUploading} className="btn cancel" onClick={this.cancelUpload}>{this.props.cancelButton && "cancel"}
                </button>
            </label>;
            else cancelButton = this.props.cancelButton
        }

        let pauseButton = null;
        if (this.props.pauseButton) {
            if (typeof this.props.pauseButton ===  "string" || typeof this.props.pauseButton ===  "boolean") pauseButton = <label>
                <button disabled={!this.state.isUploading} className="btn pause" onClick={this.pauseUpload}>{this.props.pauseButton && "pause"}
                </button>
            </label>;
            else pauseButton = this.props.pauseButton
        }

        return (
            <div id={this.props.dropTargetID} ref={node => this.dropZone = node}>
                {previousText}
                <label className={this.props.disableInput ? 'btn file-upload disabled' : 'btn file-upload'}>{textLabel}
                    <input
                        ref={node=> this.uploader = node}
                        type="file"
                        id={this.props.uploaderID}
                        className='btn'
                        name={this.props.uploaderID + '-upload'}
                        accept={this.props.fileAccept || '*'}
                        disabled={this.props.disableInput || false}
                    />
                </label>
                <div className="progress" style={{display: this.state.progressBar === 0 ? "none" : "block"}}>
                    <div className="progress-bar" style={{width: this.state.progressBar + '%'}}></div>
                </div>

                {fileList}
                {startButton}
                {pauseButton}
                {cancelButton}
            </div>
        );
    }
}

ReactResumableJs.defaultProps = {
    maxFiles: undefined,
    uploaderID: 'default-resumable-uploader',
    dropTargetID: 'dropTarget',
    filetypes: [],
    fileAccept: '*',
    maxFileSize: 10240000,
    showFileList: true,
    onUploadErrorCallback: (file, errorCount) => {
        console.log('error', file, errorCount);
    },
    onFileRemoved: function (file) {
        return file;
    },
    onCancelUpload: function () {
        return true;
    },
    onPauseUpload: function () {
        return true;
    },
    onResumeUpload: function () {
        return true;
    },
    onStartUpload: function () {
        return true;
    },
    disableDragAndDrop: false,
    fileNameServer: "",
    tmpDir: "",
    chunkSize: 1024 * 1024,
    simultaneousUploads: 1,
    fileParameterName: 'file',
    generateUniqueIdentifier: null,
    maxFilesErrorCallback: null,
    cancelButton: false,
    pause: false,
    startButton: null,
    pauseButton: null,
    previousText: "",
    headerObject : {},
    forceChunkSize: false
};
